# AfterWork Data Science: Data Wrangling with Python Project

## 1. Defining the Question

### a) Specifying the Data Analysis Question

Specify the reasearch question that you'll be answering.

> i.e. Which was the most improved country year 2017 - 2018 as per the GDP ranking?

### b) Defining the Metric for Success

We are able to determine the what exactly leads to the breakdown of the buses and provide recommendations on the same

### c) Understanding the context 

Provide recommendations on how to reduce the cases of breakdown of buses in the city of New York.

### d) Recording the Experimental Design

1. Define the Research Question
2. Data Importation
3. Data Exploration
4. Data Cleaning
5. Data Preparation
6. Data Analysis

### e) Data Relevance

The dataset is appropriate to answering the research question.

## 2. Reading the Data

In [64]:
# Importing our libraries
# ---
# Here we will import the necessary pandas libraries for data manupilation
import pandas as pd

# Here we will import the necessary numpy libraries for scientific computations
import numpy as np

# Mean imputation replaces missing values with the mean value of that feature/variable. 
from sklearn.impute import SimpleImputer

In [65]:
# Load the data below
# --- 
# Dataset url = https://bit.ly/BusBreakdownDataset
# --- 
# 
df = pd.read_csv('https://bit.ly/BusBreakdownDataset')
df.head()

School_Year  Busbreakdown_ID  ... Breakdown_or_Running_Late School_Age_or_PreK
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

In [66]:
# Checking the last 5 rows of data
# ---
df.tail()

School_Year  ...  School_Age_or_PreK
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[5 rows x 21 columns]

In [67]:
# Sample 10 rows of data
# ---
df.sample(10)

School_Year  ...  School_Age_or_PreK
187054   2017-2018  ...          School-Age
246061   2017-2018  ...          School-Age
268152   2018-2019  ...          School-Age
214156   2015-2016  ...          School-Age
223211   2017-2018  ...          School-Age
118067   2017-2018  ...          School-Age
113488   2017-2018  ...          School-Age
214570   2016-2017  ...          School-Age
58163    2016-2017  ...          School-Age
180419   2018-2019  ...          School-Age

[10 rows x 21 columns]

In [68]:
# Checking number of rows and columns
# ---
df.shape

(281110, 21)

In [69]:
# Checking datatypes
# ---
# 
df.dtypes

School_Year                        object
Busbreakdown_ID                     int64
Run_Type                           object
Bus_No                             object
Route_Number                       object
Reason                             object
Schools_Serviced                   object
Occurred_On                        object
Created_On                         object
Boro                               object
Bus_Company_Name                   object
How_Long_Delayed                   object
Number_Of_Students_On_The_Bus       int64
Has_Contractor_Notified_Schools    object
Has_Contractor_Notified_Parents    object
Have_You_Alerted_OPT               object
Informed_On                        object
Incident_Number                    object
Last_Updated_On                    object
Breakdown_or_Running_Late          object
School_Age_or_PreK                 object
dtype: object

*   Observation 1:
Most of the variables are objects i.e str data types
*   Observation 2: There was a total of 281110 records/instances that were recorded



## 3. External Data Source Validation

According to https://www.nycgo.com/neighborhoods-boroughs/about-nyc-five-boroughs/, New York City has 5 main Boroughs namely: the Bronx, Brooklyn, Manhattan, Queens and Staten Island—each with dozens of neighborhoods lending their own local flavor

Our data set contains these boroughs confirmed using the formular df.Boro.Unique(): New Jersey', 'Manhattan', 'Bronx', 'Westchester', 'Brooklyn', 'Rockland County', 'Nassau County', 'Queens', 'Staten Island', 'Connecticut', 'All Boroughs' which is more than what the information above has indicated although the data contains the 5 main boroughs. The assumption is that the extra boroughs are within the 5 main boroughs.

The data is provided to the NYC open data by Department of Education (DOE) and can be accessed via https://data.cityofnewyork.us/Transportation/Bus-Breakdown-and-Delays/ez4e-fazm

## 4. Data Preparation

### Performing Data Cleaning

In [70]:
# Checking datatypes and missing entries of all the variables
# ---
# 
df.isnull().mean() * 100

School_Year                         0.000000
Busbreakdown_ID                     0.000000
Run_Type                            0.001067
Bus_No                              0.003202
Route_Number                        0.002490
Reason                              0.000711
Schools_Serviced                    0.002490
Occurred_On                         0.000000
Created_On                          0.000000
Boro                                4.788517
Bus_Company_Name                    0.000000
How_Long_Delayed                   12.666928
Number_Of_Students_On_The_Bus       0.000000
Has_Contractor_Notified_Schools     0.000000
Has_Contractor_Notified_Parents     0.000000
Have_You_Alerted_OPT                0.000000
Informed_On                         0.000000
Incident_Number                    96.626587
Last_Updated_On                     0.000000
Breakdown_or_Running_Late           0.000000
School_Age_or_PreK                  0.000000
dtype: float64

We observe the following from our dataset:

*   Observation 1: 8 columns contain missing values while the other 13 columns have no missing values.
*   Observation 2: Boro (4%), how_long_delayed (12.66%) and incident_number (96.62%) contain the highest percentage of missing values



In [71]:
# Standardizing your dataset i.e. variable renaming
# ---
df.columns = df.columns.str.lower()

df.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

We observe the following from our dataset:

*   Observation 1: We renamed all columns to have lower cases and checked to confirm the changes. All columns now have lower cases in the column names.
*   Observation 2: Some Column names have "-" instead of white spaces which is perfect for working with the query() function



In [72]:
# Checking how many duplicate rows are there in the data
# ---
# 
df_duplicates = df[df.duplicated()]

sum(df.duplicated())

0

We observe the following from our dataset:

*   There are no duplicates in our data



In [73]:
# Checking if any of the columns are all null
# ---

df.isnull().all(axis = 1).sum()

0

We observe the following from our dataset:

*   None of the columns contains all null values



In [74]:
# Checking if any of the rows are all null
# ---
#
df.isnull().all(axis = 0).sum()

0

We observe the following from our dataset:

*   None of the rows are completely comprised of null values only.



In [75]:
# Checking if the "Yes/No" fields contain only these 2 values
# for have_you_alerted_opt variable

df['have_you_alerted_opt'].unique()

array(['Yes', 'No'], dtype=object)

We observe the following from our dataset:

*   The column "have_you_alerted_opt" contains only 2 values - Yes and No



In [76]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_parents variable
# ---
df['has_contractor_notified_parents'].unique()


array(['No', 'Yes'], dtype=object)

We observe the following from our dataset:

*   The column "has_contractor_notified_parent" contains only 2 values - Yes and No



In [77]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_schools variable
# ---

df['has_contractor_notified_schools'].unique()

array(['Yes', 'No'], dtype=object)

We observe the following from our dataset:

*   The column "has_contractor_notified_schools" contains only 2 values - Yes and No



In [78]:
# Checking unique values in break_down_or_running_late variable to ensure there is no duplication
# ---

df['breakdown_or_running_late'].unique()

array(['Running Late', 'Breakdown'], dtype=object)

In [79]:
# Checking unique values in school_age_or_prek variable
# ---

df['school_age_or_prek'].unique()

array(['School-Age', 'Pre-K'], dtype=object)

In [80]:
# Checking unique values in school_year variable
# ---

df['school_year'].unique()

array(['2015-2016', '2017-2018', '2018-2019', '2016-2017', '2019-2020'],
      dtype=object)

In [81]:
# Checking unique values in reason variable
# ---

df['reason'].unique()

array(['Heavy Traffic', 'Flat Tire', 'Other', 'Won`t Start',
       'Mechanical Problem', 'Problem Run', 'Accident',
       'Late return from Field Trip', 'Delayed by School',
       'Weather Conditions', nan], dtype=object)

In [82]:
# Checking unique values in run_type variable
# ---

df['run_type'].unique()

array(['Special Ed AM Run', 'Pre-K/EI', 'General Ed AM Run',
       'General Ed Field Trip', 'Special Ed PM Run', 'General Ed PM Run',
       'Project Read PM Run', 'Special Ed Field Trip',
       'Project Read AM Run', 'Project Read Field Trip', nan],
      dtype=object)

In [83]:
# Checking unique values in boro variable
# ---
 
df['boro'].unique() 

array(['New Jersey', 'Manhattan', 'Bronx', 'Westchester', 'Brooklyn',
       'Rockland County', 'Nassau County', nan, 'Queens', 'Staten Island',
       'Connecticut', 'All Boroughs'], dtype=object)

In [84]:
# Checking unique values in how_long_delayed variable
# ---

df.how_long_delayed.unique()

array([nan, '20MNS', '15MIN', ..., '40 -50 MIN', '80 MIN', '45 MINIUTE'],
      dtype=object)

### Overall Data Cleaning Observations
**Missing Values**

- There are a large number of missing values in the fields "How_Long_Delayed" which is important to our analysis. 
- There is an extremely large number of missing values in the "Incident_number" field but this is not incidental to our analysis and cannot be filled in without additional information.

**Error in values**

- "How_Long_Delayed" contains string values such as "MINS" or "mins" and a range of values, which needs to be changed to single integer value for our analysis.

**Error in Datatypes**

- "How_Long_Delayed" is a string datatype, should be converted to integer type.

**Error in field names**
- The column name "Boro" should be renamed to "Borough".


### Next Steps: Data Cleaning Steps

**Error in values**

- Extract the first integer value (lowest delay time) in the column "How_Long_Delayed"


**Missing Values**

- Impute the missing values in the field "How_Long_Delayed" with the mean value.


**Error in Datatypes**

- Convert "How_Long_Delayed" to int datatype.



**Error in field names**

- Rename the column "Boro" to "Borough".

In [85]:
# Lets first start by creating a copy of our dataframe 
# df_clean = df.copy(). We will use this copy as our cleaning copy.
# ---
#
df_clean = df.copy()
df_clean.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

In [86]:
# Then extracting the lowest delay time in the column how_long_delayed from the string
# 

df_clean['how_long_delayed'] = df_clean['how_long_delayed'].str.extract('(\d+)')
df_clean['how_long_delayed'].head()

0    NaN
1     20
2     15
3     15
4    NaN
Name: how_long_delayed, dtype: object

We impute the null values in 'how_long_delayed' column with mean of the column. This will take a couple of steps...

In [87]:
# We first convert our how_long_delayed to float type to allow for imputation
# ---
# 
df_clean['how_long_delayed']= df_clean['how_long_delayed'].apply(np.float64)
df_clean['how_long_delayed'].dtypes

dtype('float64')

In [88]:
# Then later perform our mean imputation 
# ---

df_clean['how_long_delayed'] = df_clean['how_long_delayed'].fillna(df_clean['how_long_delayed'].mean())
df_clean['how_long_delayed'].isnull().any()

False

In [89]:
# Then convert back our how_long_delayed column to integer datatype
# ---
# 
df_clean['how_long_delayed'] = df_clean['how_long_delayed'].apply(np.int64)
df_clean['how_long_delayed'].dtypes

dtype('int64')

In [90]:
# Then check for nulls in the column
# ---
# 
df_clean['how_long_delayed'].isnull().sum()

0

In [91]:
# Rename Boro column to Borough
# ---
# 
df_clean.columns = df_clean.columns.str.replace('boro', 'Borough')
df_clean.columns

Index(['school_year', 'busbreakdown_id', 'run_type', 'bus_no', 'route_number',
       'reason', 'schools_serviced', 'occurred_on', 'created_on', 'Borough',
       'bus_company_name', 'how_long_delayed', 'number_of_students_on_the_bus',
       'has_contractor_notified_schools', 'has_contractor_notified_parents',
       'have_you_alerted_opt', 'informed_on', 'incident_number',
       'last_updated_on', 'breakdown_or_running_late', 'school_age_or_prek'],
      dtype='object')

In [92]:
# Lastly we convert all values in our colums to lower case
# for ease of reading
# ---
# 
df_clean.columns = df.columns.str.lower()
df_clean.columns

Index(['school_year', 'busbreakdown_id', 'run_type', 'bus_no', 'route_number',
       'reason', 'schools_serviced', 'occurred_on', 'created_on', 'boro',
       'bus_company_name', 'how_long_delayed', 'number_of_students_on_the_bus',
       'has_contractor_notified_schools', 'has_contractor_notified_parents',
       'have_you_alerted_opt', 'informed_on', 'incident_number',
       'last_updated_on', 'breakdown_or_running_late', 'school_age_or_prek'],
      dtype='object')

In [93]:
# Check the first 5 record the cleaned dataset
# ---
#
df_clean.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]


## 5. Solution Implementation

Here we investigate the questions that would help craft our recommendations.

### 5.a) Questions

In [94]:
# 1. Which bus companies that had the highest breakdowns?
# ---
#

breakdowns = df_clean.groupby(['bus_company_name', 'breakdown_or_running_late'])[['busbreakdown_id']].count().reset_index()

# Sort to get bus company with highest breakdowns

breakdowns.sort_values(by = 'busbreakdown_id', ascending=0).head(15)

bus_company_name  ... busbreakdown_id
70                          G.V.C., LTD.  ...           19252
99    LEESEL TRANSPORTATION CORP (B2192)  ...           17013
149            PIONEER TRANSPORTATION CO  ...           12011
151          PIONEER TRANSPORTATION CORP  ...           11873
165            RELIANT TRANS, INC. (B232  ...           11678
95             LEESEL TRANSP CORP (B2192  ...           11382
26                    BORO TRANSIT, INC.  ...           10470
139            NEW DAWN TRANSIT, LLC (B2  ...           10292
97            LEESEL TRANSP CORP (B2192)  ...            9691
203                VAN TRANS LLC (B2192)  ...            8430
169  RELIANT TRANSPORTATION, INC (B2321)  ...            7003
183                          SNT BUS INC  ...            5993
78             HOYT TRANSPORTATION CORP.  ...            5859
88                       L & M BUS CORP.  ...            5549
104            LITTLE RICHIE BUS SERVICE  ...            5389

[15 rows x 3 columns]

In [95]:
# 2. What were the top 3 reasons for bus delays?
# ---
#

bus_delay = df_clean.groupby(['reason'])[['how_long_delayed']].count().reset_index()

# Sort to get the most frequent reason

bus_delay.sort_values(by = 'how_long_delayed', ascending = 0).head(3)

reason  how_long_delayed
3       Heavy Traffic            173221
6               Other             37579
5  Mechanical Problem             28162

In [96]:
# 3. How many students were in the buses when they broke down? 
# ---
#

number_students = df_clean.groupby(['breakdown_or_running_late'])[['number_of_students_on_the_bus']].sum().reset_index()
number_students.head()

breakdown_or_running_late  number_of_students_on_the_bus
0                 Breakdown                          58329
1              Running Late                         890829

In [97]:
# 4. Which were most frequent reasons for bus breakdowns?
# ---
#

breakdown_reasons = df_clean.groupby(['breakdown_or_running_late', 'reason'])[['bus_no']].count().reset_index()

# Sort to get most frequent reasons
# ---

breakdown_reasons.sort_values(by='bus_no', ascending=0).head()

breakdown_or_running_late              reason  bus_no
13              Running Late       Heavy Traffic  172798
16              Running Late               Other   34046
5                  Breakdown  Mechanical Problem   14985
15              Running Late  Mechanical Problem   13177
9                  Breakdown         Won`t Start    7731

In [98]:
# 5. What were the most frequent reasons for the bus running late? 
# ---
#
reasons = df_clean.groupby(['reason','how_long_delayed']).count()

# Get the records with running late reasons and sort to get most frequent reasons
# ---

reasons.sort_values(by='reason', ascending=0)

school_year  ...  school_age_or_prek
reason      how_long_delayed               ...                    
Won`t Start 5594                        1  ...                   1
            32                          1  ...                   1
            0                         154  ...                 154
            1                         178  ...                 178
            2                           1  ...                   1
...                                   ...  ...                 ...
Accident    120                         6  ...                   6
            454                         2  ...                   2
            1003                        1  ...                   1
            2308                        1  ...                   1
            0                          32  ...                  32

[491 rows x 19 columns]

In [99]:
# 6. What was the average delay time for each reason type?
# ---
# 
avg_delay = df_clean.groupby('reason')[['how_long_delayed']].mean().reset_index()

# Get the records with reasons and how long on average a delay took then sorted the records
# ---

avg_delay.sort_values(by='how_long_delayed', ascending=0)

reason  how_long_delayed
4  Late return from Field Trip         62.235975
0                     Accident         34.965162
7                  Problem Run         31.136676
5           Mechanical Problem         30.915205
9                  Won`t Start         29.644061
2                    Flat Tire         29.392343
8           Weather Conditions         29.227351
6                        Other         28.421006
3                Heavy Traffic         26.225550
1            Delayed by School         18.907741

### 5.b) Recommendations

From the above analysis, below are our recommendations:

1.   Since most of the delays were caused by Late return from Field Trips, it is recommended that the buses be rotated out to prevent breakdown i.e the buses going out for field trips should be on rotational basis.
2.   Heavy traffic is the most frequent reason for running late and hence either the roads be expanded to ease traffic or the number of vehicles plying that route should be regulated. Alternative routes may also be considered.
3. An investigation into LITTLE RICHIE BUS SERVICE which is the bus company with the most breakdowns should be launched to understand the reason.





## 6. Challenging your Solution

During this step, we review our solution and implement  approaches that could potentially provide a better outcome. In our case, we could propose the following question that wasn't answered in our solution because it couldn't have greatly contributed to our recommendation.

In [102]:
# Which boroughs experienced the most breakdowns?
# ---  
#
breakdowns = df_clean.groupby(['boro', 'breakdown_or_running_late'])[['bus_no']].count().reset_index()

# Sort to get borough with highest breakdowns


breakdowns.sort_values(by = 'bus_no', ascending=0)

boro breakdown_or_running_late  bus_no
9         Manhattan              Running Late   64817
3             Bronx              Running Late   61662
5          Brooklyn              Running Late   55412
15           Queens              Running Late   28938
19    Staten Island              Running Late   13453
2             Bronx                 Breakdown    9287
14           Queens                 Breakdown    9118
21      Westchester              Running Late    6954
4          Brooklyn                 Breakdown    6826
11    Nassau County              Running Late    3439
8         Manhattan                 Breakdown    3359
13       New Jersey              Running Late    1517
17  Rockland County              Running Late     833
10    Nassau County                 Breakdown     685
18    Staten Island                 Breakdown     454
1      All Boroughs              Running Late     409
7       Connecticut              Running Late     200
20      Westchester                 Breakdown     139
12       New Jersey                 Breakdown      66
16  Rockland County                 Breakdown      42
0      All Boroughs                 Breakdown      20
6       Connecticut                 Breakdown      10

Our observations:

1.   Manhattan and Bronx experienced the most number of breakdowns.
2.  Connecticut experienced the least number of breakdowns.


How does this observation tie to our solution?



## 7. Follow up questions

During this step, you rethink and propose other ways that you can improve your solution.

### a). Did we have the right data?

Yes we did


### b). Do we need other data to answer our question?

The data available was sufficient to answer the questions.

### c). Did we have the right question?

We answered the right questions